### ISSUE TITLE
Implement immutable Vector, HashMap and HashSet backed by hash tries

### ISSUE THREAD

**Issue by orionll** [functionaljava/functionaljava#39]
While red-black trees are suitable for ordered elements, we also need data structures which can hold elements with no notion of ordering.
We should implement 3 types of immutable collections:
- Vector - sequence of elements
- HashSet - set of elements (no repetitions)
- HashMap - map

These collections can be implemented using hash tries (like in Scala)

**Comment by mrbackend** [comment]
I'm thinking it would be fun having a go at this one during Christmas, unless there are any objections.

**Comment by mperry** [comment]
:+1:

**Comment by mrbackend** [comment]
I've started on this, but it won't make the 4.5 release.

**Comment by mperry** [comment]
No problem @mrbackend.  Do you have a target date for an initial implementation?  You probably already know this, but I believe Javaslang has implemented this too, so it might be worth looking at their implementation too.

**Comment by danieldietrich** [comment]
:-)

**Comment by danieldietrich** [comment]
@mrbackend You find [here the according Javaslang issue](https://github.com/javaslang/javaslang/issues/170). @ruslansennov did the implementation. It is based on a LAMP paper and on a Haskell implementation. It would be great if we can learn from each other!

**Comment by mrbackend** [comment]
I'm sorry, with a full time day job and family, and still measuring my FP experience in months rather than years (I have 20+ years of imperative programming, though), I'm not able to give a target date.

I didn't know about Javaslang, but I've looked it up and both they and Scala use a HAMT as a helper type for HashMap. (I think the Javaslang implementation looks beautiful!). Scala's Vector also uses a trie, but not the one used by HashMap. I have the idea, and a partial implementation, of a more general purpose 32-bit key compact trie to be used by both HashMap and Vector. It does not have any collision handling (that must be left to HashMap), but otherwise it is along the same lines as Javaslang's HAMT, except this one merges single path nodes vertically. Also, I haven't landed on 5 or 6 bits per level, I have to test performance differences first.

BTW, I need suggestions for other names than HashMap, HashSet and Vector. The first two are already taken, and we have a package named vector .

**Comment by danieldietrich** [comment]
@mrbackend thank you for the kind words! Ruslan did an awesome job implementing the HAMT! For Vector I was also interested in the 'Bitmapped Vector Trie' Daniel Spiewak talked about in his talk [Extreme Cleverness: Functional Data Structures in Scala](daniel spiewak bitmapped vector trie) (move forward to minute 32). I think Daniel implemented Scala's current Vector on this Trie. It would be super-interesting to see a Java-port and see benchmarks Bitmapped Vector Trie vs. HAMT.

**Comment by mperry** [comment]
No problem @mrbackend, just wondering if you had a date in mind.  Have you push your progress up to a branch on your fork?  I could take a look a the progress so far.

**Comment by mrbackend** [comment]
@mperry There has been a lot of back and forth between doing the implementation FP-idiomatically correct and doing it more traditionally Java-like (still being pure, though). Java's type inference isn't very good, especially for curried functions, so I have been spending more than 50% of the time getting it to compile with the least amount of type specifiers. Besides, while seasoned FP'ers might have been able to comprehend my attempts at idiomatic code, I am at a point where I don't understand my own (working) code. I'll tidy it up the next weekend and push it then.

**Comment by aphexcx** [comment]
Eagerly awaiting immutable maps - may I suggest putting these under an `.immutable` package, like Scala?

Also, will operations such as `.filter` be included?

**Comment by mperry** [comment]
@aphexcx, note that there is already an immutable map, fj.data.TreeMap backed by a red-black tree.  To use the map, you need to provide an ordering, `Ord<K>`, for the keys.

**Comment by mperry** [comment]
@mrbackend Did you end up pushing to your local repo?  I am tempted to go in and do this, but don't want to duplicate your effort.

**Comment by clinuxrulz** [comment]
Maybe I can offer some inspiration: https://github.com/DanielGronau/highj/blob/master/src/main/java/org/highj/data/IntMap.java

Its a implementation of a persistent IntMap, and I believe it works very similar to the HashMaps in Clojure.

With ```IntMap``` you can of course easily define a persistent ```HashMap``` by using ```Hash<A>``` to generate the integer keys. And from ```HashMap``` you can derive ```HashSet``` by going the ```HashSet<A> = HashMap<A,Unit>``` route. It does not solve the ```Vector``` issue, but comes close. ( Using ```IntMap``` to make a ```Vector``` you end up with gaps, making it really sparse vector).

**Comment by clinuxrulz** [comment]
Also ```HashMap<K,V> = IntMap<List<V>>``` when given ```Hash<K>``` and ```Eq<K>``` so you can handle collisions in the hash keys.

**Comment by clinuxrulz** [comment]
Actually ```IntMap``` might not have good performance with full 32-bit key sizes.

**Comment by mrbackend** [comment]
TL; DR: I'm on it.

This one's been left in a poor state for a couple of months, right now it doesn't even compile. I already use something like IntMap (my name: IntKeyedTrie), which is a more general one that (hopefully) can be used by Vector as well. Most of my work so far has been on IntKeyedTrie, so we will need a discussion on the "ImmutableHashMap" and "ImmutableVector" (name suggestions welcome) APIs later. Using "ImmutableHashMap" to implement "ImmutableHashSet" is a given, I think.

When I temporarily left it, I was experimenting with different branch node implementations to speed it up. A packed/sparse array node implementation is mandatory not to waste too much space, but at some point (25-50% full nodes ?), a packed array is just wasting CPU. My IntKeyedTrie needs implementations of both packed and direct access/fast array nodes to reach the same performance ballpark as Javaslang/Scala. Still experimenting, in other words.

I am also experimenting with the number of bits per level, which is configurable (at least for now). Also, diffusing hash values may be an issue (to ensure that they are evenly spread all over the 32-bit range; locality is an issue for Vector, but not for HashMap), but at the moment I think the best alternative is to create a special instance of Hash<Integer> for this (I've implemented one using FNV-1a), and just mention it in the Javadoc.

To summarize: I need to stop fiddling around, and start making a stable version that just works.

@clinuxrulz Why to you think IntMap might not have good performance with full 32-bit key sizes?

**Comment by clinuxrulz** [comment]
Because I've had a little bad luck with it performance wise. There has been an attempt to derive a HashMap for an IntMap in highj. DanielGronau/highj#37

The version we are competing with in the naive unbalanced binary tree. And the naive version outperformed the IntMap based one straight away, until I tried to mask the hash value produced by the key to keep the lookup key small. Only then did I slightly outperformed the naive version, but not by enough to be worth it.

Since the keys are random, the naive version tends to be close to balanced, which helped it a bit too I think.

**Comment by dobesv** [comment]
I don't think we need to use the word "Immutable" in class names in FJ - e.g. `List` rather than `ImmutableList`.  If `HashMap` seems too short maybe something that speaks more to its structure, like `VectorHashMap` or `FingerHashMap`.  Sorry, those are pretty bad... hopefully it sparks a better idea.

**Comment by mperry** [comment]
@mrbackend Have you pushed to a repo somehwere?  I had a go at this myself, see https://github.com/mperry/hash-array-mapped-trie.  Still incomplete and there were a few things I was not sure of:
- I use the existing Seq implement, rather than arrays.  This means faster writes than arrays, but slower reads.
- I had to implement an immutable BitSet which probably isn't done very well.  Looking at other libs would surely make this better.

**Comment by mperry** [comment]
I have pushed these changes up to my FunctionalJava fork, see https://github.com/mperry/functionaljava/tree/hamt.

**Comment by mperry** [comment]
Changes are ready to be merged.  The BitSet implementation is not great, but it is probably good enough.  See #280.
